### ISSUE TITLE
采用  --loss-type = 4 的时候 acc=0，lossvalue始终很小，epoch也已经达到20了，请问是怎么回事

### ISSUE THREAD

**Issue by NOON47** [deepinsight/insightface#783]
采用  --loss-type = 4 的时候 acc=0，lossvalue始终很小，epoch也已经达到20了，请问是怎么回事

lr-batch-epoch: 0.1 279 20
INFO:root:Epoch[20] Batch [280] Speed: 223.32 samples/sec       acc=0.000000    lossvalue=0.000012
INFO:root:Epoch[20] Batch [300] Speed: 223.86 samples/sec       acc=0.000000    lossvalue=0.000016
INFO:root:Epoch[20] Batch [320] Speed: 223.30 samples/sec       acc=0.000000    lossvalue=0.000016
INFO:root:Epoch[20] Batch [340] Speed: 223.48 samples/sec       acc=0.000000    lossvalue=0.000018
INFO:root:Epoch[20] Batch [360] Speed: 223.33 samples/sec       acc=0.000000    lossvalue=0.000010
INFO:root:Epoch[20] Batch [380] Speed: 223.15 samples/sec       acc=0.000000    lossvalue=0.000013
INFO:root:Epoch[20] Batch [400] Speed: 222.99 samples/sec       acc=0.000000    lossvalue=0.000011
INFO:root:Epoch[20] Batch [420] Speed: 223.47 samples/sec       acc=0.000000    lossvalue=0.000024
INFO:root:Epoch[20] Batch [440] Speed: 223.31 samples/sec       acc=0.000000    lossvalue=0.000085
INFO:root:Epoch[20] Batch [460] Speed: 223.44 samples/sec       acc=0.000000    lossvalue=0.000058
INFO:root:Epoch[20] Batch [480] Speed: 223.40 samples/sec       acc=0.000000    lossvalue=0.000043
INFO:root:Epoch[20] Batch [500] Speed: 223.35 samples/sec       acc=0.000000    lossvalue=0.000032

**Comment by NOON47** [comment]
@

**Comment by YokkaBear** [comment]
我训练中也存在类似的问题：  
```
INFO:root:Epoch[0] Batch [6980-7000]	Speed: 47.55 samples/sec	acc=0.000000	lossvalue=11.392895
INFO:root:Epoch[0] Batch [7000-7020]	Speed: 47.09 samples/sec	acc=0.000000	lossvalue=11.393996
INFO:root:Epoch[0] Batch [7020-7040]	Speed: 47.47 samples/sec	acc=0.000000	lossvalue=11.386449
INFO:root:Epoch[0] Batch [7040-7060]	Speed: 47.36 samples/sec	acc=0.000000	lossvalue=11.389718
INFO:root:Epoch[0] Batch [7060-7080]	Speed: 47.18 samples/sec	acc=0.000000	lossvalue=11.382548
INFO:root:Epoch[0] Batch [7080-7100]	Speed: 47.33 samples/sec	acc=0.000000	lossvalue=11.375066
INFO:root:Epoch[0] Batch [7100-7120]	Speed: 47.65 samples/sec	acc=0.000000	lossvalue=11.384797
INFO:root:Epoch[0] Batch [7120-7140]	Speed: 47.36 samples/sec	acc=0.000000	lossvalue=11.384455
INFO:root:Epoch[0] Batch [7140-7160]	Speed: 47.18 samples/sec	acc=0.000000	lossvalue=11.381091
INFO:root:Epoch[0] Batch [7160-7180]	Speed: 47.53 samples/sec	acc=0.000000	lossvalue=11.387330
INFO:root:Epoch[0] Batch [7180-7200]	Speed: 46.98 samples/sec	acc=0.000000	lossvalue=11.385892
INFO:root:Epoch[0] Batch [7200-7220]	Speed: 47.36 samples/sec	acc=0.000000	lossvalue=11.371730
INFO:root:Epoch[0] Batch [7220-7240]	Speed: 47.41 samples/sec	acc=0.000000	lossvalue=11.385509
INFO:root:Epoch[0] Batch [7240-7260]	Speed: 47.11 samples/sec	acc=0.000000	lossvalue=11.355889
INFO:root:Epoch[0] Batch [7260-7280]	Speed: 47.20 samples/sec	acc=0.000000	lossvalue=11.407780
INFO:root:Epoch[0] Batch [7280-7300]	Speed: 47.10 samples/sec	acc=0.000000	lossvalue=11.383916
INFO:root:Epoch[0] Batch [7300-7320]	Speed: 47.45 samples/sec	acc=0.000000	lossvalue=11.384495
INFO:root:Epoch[0] Batch [7320-7340]	Speed: 47.23 samples/sec	acc=0.000000	lossvalue=11.359317
INFO:root:Epoch[0] Batch [7340-7360]	Speed: 47.13 samples/sec	acc=0.000000	lossvalue=11.374713
INFO:root:Epoch[0] Batch [7360-7380]	Speed: 47.49 samples/sec	acc=0.000000	lossvalue=11.369580
INFO:root:Epoch[0] Batch [7380-7400]	Speed: 47.16 samples/sec	acc=0.000000	lossvalue=11.391267
```
虽然验证集准确率是有的，但是也很低：  
```
testing verification..
(12000, 512)
infer time 129.6471179999998
[lfw][8000]XNorm: 16.671212
[lfw][8000]Accuracy-Flip: 0.67467+-0.02408
testing verification..
(14000, 512)
infer time 151.49728699999986
[cfp_fp][8000]XNorm: 21.560114
[cfp_fp][8000]Accuracy-Flip: 0.57843+-0.01179
testing verification..
(12000, 512)
infer time 129.7342089999998
[agedb_30][8000]XNorm: 20.670939
[agedb_30][8000]Accuracy-Flip: 0.49867+-0.00433
saving 4
INFO:root:Saved checkpoint to "./models/r100-softmax-emore/model-0004.params"
[8000]Accuracy-Highest: 0.50083
```
使用的数据集是CASIA-Weface，batch_size设置是4，lr设置是1e-10（只有两块显卡而且内存很小），望up主或者训练成功的大佬指点迷津
@nttstar

**Comment by becauseofAI** [comment]
@NOON47 @YokkaBear 
After a long training period, I finally got the following results:
```
2019-07-11 08:38:16,452 Epoch[0] Batch [10980-11000]    Speed: 17.50 samples/sec        acc=0.000586    lossvalue=28.366652
2019-07-11 08:41:49,111 Epoch[0] Batch [11000-11020]    Speed: 24.08 samples/sec        acc=0.001758    lossvalue=28.055955
2019-07-11 08:45:19,096 Epoch[0] Batch [11020-11040]    Speed: 24.38 samples/sec        acc=0.000977    lossvalue=27.956807
2019-07-11 08:48:52,124 Epoch[0] Batch [11040-11060]    Speed: 24.03 samples/sec        acc=0.001563    lossvalue=28.059857
2019-07-11 08:52:18,685 Epoch[0] Batch [11060-11080]    Speed: 24.79 samples/sec        acc=0.001563    lossvalue=27.804412
```
Maybe you need to wait for a very long time patiently.

**Comment by YokkaBear** [comment]
> @NOON47 @YokkaBear
> After a long training period, I finally got the following results:
> 
> ```
> 2019-07-11 08:38:16,452 Epoch[0] Batch [10980-11000]    Speed: 17.50 samples/sec        acc=0.000586    lossvalue=28.366652
> 2019-07-11 08:41:49,111 Epoch[0] Batch [11000-11020]    Speed: 24.08 samples/sec        acc=0.001758    lossvalue=28.055955
> 2019-07-11 08:45:19,096 Epoch[0] Batch [11020-11040]    Speed: 24.38 samples/sec        acc=0.000977    lossvalue=27.956807
> 2019-07-11 08:48:52,124 Epoch[0] Batch [11040-11060]    Speed: 24.03 samples/sec        acc=0.001563    lossvalue=28.059857
> 2019-07-11 08:52:18,685 Epoch[0] Batch [11060-11080]    Speed: 24.79 samples/sec        acc=0.001563    lossvalue=27.804412
> ```
> 
> Maybe you need to wait for a very long time patiently.

thx a lot, I'll try more training epochs and see if any changes later.
BTW, could you please further explain the reason why train_acc=0 whilst verification_acc is normal? And if that's the case, whether the model can be used in real-situation face recognition or not? thx again and I'm really confused about that.

**Comment by JackWolftoMc** [comment]
> @NOON47 @YokkaBear
> After a long training period, I finally got the following results:
> 
> ```
> 2019-07-11 08:38:16,452 Epoch[0] Batch [10980-11000]    Speed: 17.50 samples/sec        acc=0.000586    lossvalue=28.366652
> 2019-07-11 08:41:49,111 Epoch[0] Batch [11000-11020]    Speed: 24.08 samples/sec        acc=0.001758    lossvalue=28.055955
> 2019-07-11 08:45:19,096 Epoch[0] Batch [11020-11040]    Speed: 24.38 samples/sec        acc=0.000977    lossvalue=27.956807
> 2019-07-11 08:48:52,124 Epoch[0] Batch [11040-11060]    Speed: 24.03 samples/sec        acc=0.001563    lossvalue=28.059857
> 2019-07-11 08:52:18,685 Epoch[0] Batch [11060-11080]    Speed: 24.79 samples/sec        acc=0.001563    lossvalue=27.804412
> ```
> 
> Maybe you need to wait for a very long time patiently.



> > @NOON47 @YokkaBear
> > After a long training period, I finally got the following results:
> > ```
> > 2019-07-11 08:38:16,452 Epoch[0] Batch [10980-11000]    Speed: 17.50 samples/sec        acc=0.000586    lossvalue=28.366652
> > 2019-07-11 08:41:49,111 Epoch[0] Batch [11000-11020]    Speed: 24.08 samples/sec        acc=0.001758    lossvalue=28.055955
> > 2019-07-11 08:45:19,096 Epoch[0] Batch [11020-11040]    Speed: 24.38 samples/sec        acc=0.000977    lossvalue=27.956807
> > 2019-07-11 08:48:52,124 Epoch[0] Batch [11040-11060]    Speed: 24.03 samples/sec        acc=0.001563    lossvalue=28.059857
> > 2019-07-11 08:52:18,685 Epoch[0] Batch [11060-11080]    Speed: 24.79 samples/sec        acc=0.001563    lossvalue=27.804412
> > ```
> > 
> > 
> > Maybe you need to wait for a very long time patiently.
> 
> thx a lot, I'll try more training epochs and see if any changes later.
> BTW, could you please further explain the reason why train_acc=0 whilst verification_acc is normal? And if that's the case, whether the model can be used in real-situation face recognition or not? thx again and I'm really confused about that.

This is so confusing that loss is very low while train_acc keeps at the level of 0 (though I think it should be 1). I encountered a different situation where loss from zero enhanced to nan suddenly as the acc from 1 to 0. Some said I need to make changes to parameters like lr or batch_size.

**Comment by NOON47** [comment]
> @NOON47 @YokkaBear
> After a long training period, I finally got the following results:
> 
> ```
> 2019-07-11 08:38:16,452 Epoch[0] Batch [10980-11000]    Speed: 17.50 samples/sec        acc=0.000586    lossvalue=28.366652
> 2019-07-11 08:41:49,111 Epoch[0] Batch [11000-11020]    Speed: 24.08 samples/sec        acc=0.001758    lossvalue=28.055955
> 2019-07-11 08:45:19,096 Epoch[0] Batch [11020-11040]    Speed: 24.38 samples/sec        acc=0.000977    lossvalue=27.956807
> 2019-07-11 08:48:52,124 Epoch[0] Batch [11040-11060]    Speed: 24.03 samples/sec        acc=0.001563    lossvalue=28.059857
> 2019-07-11 08:52:18,685 Epoch[0] Batch [11060-11080]    Speed: 24.79 samples/sec        acc=0.001563    lossvalue=27.804412
> ```
> 
> Maybe you need to wait for a very long time patiently.

hello,did you use the  parameter width --loss-type=4 and ----pretrained ‘./model-y1-test2/model,0'？i used these parameters to train my dataset,but wait for a long time my acc still 0,but val is not zero.
lr-batch-epoch: 0.0001 415 437
testing verification..
(12000, 128)
infer time 28.587003
[lfw][364000]XNorm: 36.869989
[lfw][364000]Accuracy-Flip: 0.83883+-0.02060
testing verification..
(41044, 128)
infer time 95.989467
[mydata][364000]XNorm: 39.555985
[mydata][364000]Accuracy-Flip: 0.90137+-0.02584
saving 182
INFO:root:Saved checkpoint to "./models/mobilefacenet_lr0_001/model_y1_arcface-0182.params"
[364000]Accuracy-Highest: 0.91375
INFO:root:Epoch[437] Batch [420]        Speed: 19.00 samples/sec        acc=0.000000    lossvalue=0.000028
INFO:root:Epoch[437] Batch [440]        Speed: 228.14 samples/sec       acc=0.000000    lossvalue=0.000033
INFO:root:Epoch[437] Batch [460]        Speed: 225.99 samples/sec       acc=0.000000    lossvalue=0.000030
INFO:root:Epoch[437] Batch [480]        Speed: 226.39 samples/sec       acc=0.000000    lossvalue=0.000033
INFO:root:Epoch[437] Batch [500]        Speed: 226.56 samples/sec       acc=0.000000    lossvalue=0.000026
INFO:root:Epoch[437] Batch [520]        Speed: 226.10 samples/sec       acc=0.000000    lossvalue=0.000028
INFO:root:Epoch[437] Batch [540]        Speed: 226.20 samples/sec       acc=0.000000    lossvalue=0.000038
INFO:root:Epoch[437] Batch [560]        Speed: 226.90 samples/sec       acc=0.000000    lossvalue=0.000032
INFO:root:Epoch[437] Batch [580]        Speed: 226.26 samples/sec       acc=0.000000    lossvalue=0.000029

**Comment by WIll-Xu35** [comment]
@NOON47 Did you modify the dataset? I encountered this problem while using my own dataset. In config.py, I entered the wrong value of num_classes for my own dataset. And I got the exact problem, acc = 0 and very small lossvalue.

Maybe you can check your parameters for your training dataset and see if all values are set correctly.

**Comment by NOON47** [comment]
@WIll-Xu35  Hello ,i have modified my dataset ,and wo got de pre-train model:MobileFaceNet,ArcFace@ms1m-refine-v1 .Now i use this model to train my dataset.

**Comment by WIll-Xu35** [comment]
@NOON47 How many ids are in your prepared dataset? In config.py, does the num_classes value match your id count? I encountered this problem because I set the wrong value for num_classes.

**Comment by S130111** [comment]
@NOON47 hi 兄弟 问题解决了吗？我也遇到这样的情况，您是怎么解决的？能跟我讲解一下吗？十分感谢
